text/0000-builder-naming.md

```diff
@@ -0,0 +1,83 @@
+- Start Date: 2015-04-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Provide guidelines for builder types in the standard library.
+
+# Motivation
+
+We use builder pattern in a few places in the standard library, but they are not consistent. We
+have `thread::Builder`, `process::Command` and there’s `fs::OpenOptions`. They are somewhat ad-hoc
+and inconsistency makes forming the intuition about them hard.
+
+A common example is people not knowing about `OpenOptions` builder and being confused by its name
+and lack of discoverability.
+
+# Detailed design
+
+## Naming
+
+If the module containing the builder is small and module name represents itself clearly (such as
+`thread` or `process`), the builder should be exported as `Builder`. For example:
+
+* `thread::Builder` builds and spawns threads;
+* `process::Builder` builds and spawns processes;
+
+Otherwise, the name of the built object is suffixed with `Builder` to make it clear this is a
+builder and what object can be expected as an outcome. For example, `fs::OpenOptions`, if it
+followed this guideline, would be `fs::FileBuilder` instead, because the final result is `fs::File`
+object.
+
+Otherwise, if there’s no clear object that can be built using the builder (most likely because it
+contains a several different finalizers), a noun describing all the buildable objects and suffixed
+with the word `Builder` is used. For example: `ThreadBuilder`, even though it produces
+`JoinHandle`s and `JoinGuard`s, rather than `Thread`s.
+
+## Methods
+
+Option setters – methods which represent options that can be set on the builder – should receive
+the builder object as `&mut self` rather than `self` unless there is a good reason to do otherwise.
+
+Finalizers – methods which produce the built object – should consume the builder object (i.e.
+receive `self`, rather than a reference), unless there is a good reason to do otherwise.
```

Comment:
I think that this finalizer convention actually needs to be the same as the above convention which is to say that `Foo::new().bar().finalize()` won't actually work if it consumes `self` (as `bar()` returned `&mut self`), perhaps this could recommend "a borrowed self" instead?